### ISSUE TITLE
StorageId: Dataset storage id for Swift/S3 is not accurate until publish

### ISSUE THREAD

**Issue by ferrys** [IQSS/dataverse#4460]
The storage identifier for a swift based-dataset is set as `swift://10.5072/FK2/NEXT0C` pre-publish and `swift://endpoint1:doi-10-5072-FK2-NEXT0C` post publish, which is what it should be the entire time. 

This is because in `CreateDatasetCommand`, a new DataAccess object is never created and therefore the _draft_ dataset never goes through the `StorageIO` process of being created. My proposed solution for this is to remove the code for setting a storage identifier in `CreateDatasetCommand` and instead call `DataAccess.createNewStorageIO(theDataset,"dataset")`, which temporarily sets the dataset storage identifier to "dataset" and then forces the dataset to go through the process of getting its storage identifier reset by one of the AccessIO classes.

**Comment by ferrys** [comment]
Interesting to note that the storage identifier for local storage actually does the opposite: pre-publish it is `file://10.5072/FK2/NEXT0C` and post-publish it is just `file`.

**Comment by landreev** [comment]
Yes, the name of the bucket (as in "endpoint1" in "swift://endpoint1:doi-10-5072-FK2-NEXT0C") should definitely be there, for the storageidentifier to be meaningful. (because there may be multiple end points/S3 buckets defined on the system). So yes, insuring that it's there from the get go is the right thing to do. 
That said, I never understood why we had the "doi-10-5072-FK2-NEXT0C" in that dataset-level storage identifier... Because it's something that can be derived in real time from the persistent identifier of the dataset. Like, for files all we store in the storageidentifier is "swift://endpoint1:fileidentifier" - and the corresponding folder name is derived in real time from the persistent id of the parent dataset... So it would kind of make sense to just store "swift://endpoint1" for the dataset... 
(Not insisting on having it changed as part of this same issue though)

**Comment by landreev** [comment]
As for the fix to the exports not working with S3... I'm now realizing that .open() call was being skipped for a reason... We were trying to optimize S3 for a case of a dataset page full of thumbnails; S3 is like the worst way for this use case, of retrieving large numbers of tiny files. (it's optimized for storing large files; and lookups on file metadata are often more expensive than reading a thumbnail-size file...) So the whole point was to be able to just go to the bytes, without doing any lookups... So we have 
```
if (dvObject instanceof DataFile) {
            return getMainFileKey() + "." + auxItemTag;
} 
```
there, and in getMainFileKey() there is code for generating the key if it's null... But then if it's a daset, it just expects the key not to be null: 
```
else if (dvObject instanceof Dataset) {
            return key + "/" + auxItemTag;
}
```
in other words, we just missed this case... 
So, instead of just putting that .open() there, it should only be executed inside the "else if (dvObject instanceof Dataset)" clause. Or maybe some code for creating a key on the cheap,  without the overhead of a full "open". Similarly to what we are doing for files there...
OK, I'm willing to go there and change this part myself. But, once again, thanks for pointing out that it was broken. That was kinda embarrassing...

**Comment by landreev** [comment]
(I checked in 2 lines last night, changing your fix just so that dataset-level aux files are still working, but we can still access thumbnails and other datafile-level aux files without calling a full open()...)

**Comment by ferrys** [comment]
> how does setting the storageidentifier to "dataset" solve this (or works at all)? Where and how is it going to be forced to be reset to a real storage id?

@landreev and I chatted about this and decided that setting the storageIdentifier to "dataset" was confusing and have instead set it to "placeholder" in all instances where `createNewStorageIO` is called with a DvObject that is a dataset.

By calling `DataAccess.createNewStorageIO(dataset, "placeholder")`, we are passing "placeholder" as a temporary storageIdentifier for the dataset. It exists as a placeholder only until it is reset when `storageIO.open(DataAccessOption.WRITE_ACCESS);` is called on line 108 of `DataAccess.java`.
